Configurable mapname

superkhau opened the pull request:
Connect to https://github.com/strongloop-internal/scrum-loopback/issues/1173

cc @bajtos 

# Prerequisites

- [x] [Renaming `flush` to deleteAll` in Juggler](https://github.com/strongloop/loopback-datasource-juggler/pull/1214)

superkhau commented:
@slnode test please

superkhau commented:
@slnode test please

superkhau commented:
@bajtos PTAL, this should be ready to land.

superkhau commented:
@slnode test please

superkhau commented:
@slnode test please

superkhau commented:
@rmg I need to reconfigure the env var passed in by CI to not use a default mapname (the last portion of the url -- ie `/...`) to test this feature. Can you point me in the right direction to fix that?

bajtos commented:
>  I need to reconfigure the env var passed in by CI to not use a default mapname

See https://github.ibm.com/apimesh/ci/blob/4c762455c3066f2d6e316d4f4c07847103e9acd5/common/CI.groovy#L57

For short term, I think it may be ok to change dataSource setup in the tests to remove trailing `/default` from the url provide via ENV var. Something like

```js
if (/\/wxsdata\/v1\/grids\/[^\/]+/default$/.test(url))
  url = url.replace(/\/default$/, '');
```

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -2,9 +2,12 @@
 // Node module: loopback-connector-kv-extreme-scale
 // This file is licensed under the MIT License.
 // License text available at https://opensource.org/licenses/MIT
-
```
I personally prefer to keep an empty line after the license boilerplate, I believe other files are doing the same, so it's better to preserve consistency.

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -2,9 +2,12 @@
 
+var SG = require('strong-globalize');
+SG.SetRootDir(__dirname + '/..');
+var g = SG();
```
Can we move this to `lib/globalize.js` please so that when we decide to add more files to `lib` in the future, then the globalization setup is already there?

superkhau replied:
Thanks for fixing it via https://github.com/strongloop/strong-globalize/pull/92, I remember it was broken a long time ago so I didn't bother moving it out.

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -93,35 +94,31 @@ util.inherits(ExtremeScaleKVConnector, Connector);
+      // The workaround is to delete the key first
+      this._clear(modelName, key, (err) => {
+        debug(
+          'Warning: cannot delete the key %s before setting a new value. %s',
```
please include the model name in the message

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -158,32 +154,29 @@ ExtremeScaleKVConnector.prototype._clear = function(composedKey, callback) {
 
-    this._get(composedKey, (err, body) => {
+    if (body === null) {
+      debug('GET %j -> (404 Not Found)', modelName);
```
please include key (and keep the modelName) in the message

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -275,61 +268,46 @@ function(modelName, filter, options) {
-ExtremeScaleKVConnector.prototype._query = function(query, callback) {
+ExtremeScaleKVConnector.prototype._query =
+function(modelName, query, callback) {
+  const mapname = this.mapname(modelName);
```
I believe we agreed to modify `_query` so that it does not need to call `mapname` and let `this.request` handle that part?

```js
requestOptions = {
  baseUrl: this._queryUrl,
  // ...
};
```

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -275,61 +268,46 @@ function(modelName, filter, options) {
+ExtremeScaleKVConnector.prototype.request =
+function(modelName, options, callback) {
+  const mapname = this.mapname(modelName);
+  const baseUrl = this._url + '/' + mapname;
```
`(options.baseUrl || this._url)`?

Then we will also need to fix `extend` so that `options.baseUrl` does not override this computed `baseUrl`.

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -353,6 +331,31 @@ ExtremeScaleKVConnector.prototype.request = function(options, callback) {
+    uri: '/',
+    method: 'DELETE',
+  };
+  this.request(modelName, requestOptions, callback);
```
Can you add a `debug` log please, for consistency with other KVAO methods?

bajtos reviewed (changes requested):
Reviewed changes in lib, will review remaining changes soon.

bajtos commented on test/unit/kv-extreme-scale.unit.js:
```diff
@@ -0,0 +1,73 @@
+
+    context('without user-defined mapname', () => {
+      const expectedMapname = modelName + keyExpirationDefaults;
+      before(createTestModel);
```
I find it puzzling how is `expectedName` related to the model created by `createTestModel`.

I am proposing to:
 - make `createTestModel` name more descriptive by describing what kind of model is created
 - or inline the function here, so that it's immediately obvious what kind of model is created.

bajtos commented on test/unit/kv-extreme-scale.unit.js:
```diff
@@ -0,0 +1,73 @@
+
+    context('with user-defined mapname', () => {
+      const expectedMapname = 'JavaMap' + keyExpirationDefaults;
+      before(createTestModel);
```
Same comment as above.

rmg commented:
👍  to @bajtos suggestion of making the tests accept the URL with and without the name.

```js
const env = process.env;
env.EXTREME_SCALE_URL = (env.EXTREME_SCALE_URL || '').replace(/\default$/, '');
```

simonhoibm committed:
Add globalization

simonhoibm committed:
Add configurable mapname

* Refactor code to use user-defined mapname property in model
  definition (defaults to model name if undefined)

* Do not compose keys when setting key/value pairs (also removed
  dependency on ModelKeyComposer)

* Add deleteAll operator to fix false positives from integration tests

superkhau commented:
@bajtos PTAL again

superkhau commented:
@slnode test please

superkhau commented:
@slnode test please

superkhau commented:
@bajtos PTAL, it's finally green. :shipit:

bajtos commented on lib/kv-extreme-scale.js:
```diff
@@ -322,17 +291,21 @@ ExtremeScaleKVConnector.prototype._query = function(query, callback) {
+  const mapname = this.mapname(modelName);
+  const baseUrl = (options.baseUrl || this._url) + '/' + mapname;
+
+  options = Object.assign(options, {
```
`Object.assign` modifies the first argument. As a result, the `options` argument passed to the method is modified too, which I consider evil.

Here is a better solution:

```js
options = Object.assign({}, options, {
  // ...
});
```

bajtos commented on test/helpers/data-source-factory.js:
```diff
@@ -10,8 +10,14 @@ const connector = require('../..');
+// Allow urls with hardcoded mapname or dynamic mapname based on model names
+// (ie. no trailing `/default` in the url).
+let url = process.env.EXTREME_SCALE_URL;
+if (/\/wxsdata\/v1\/grids\/[^\/]+\/default$/.test(url))
```
`process.env.EXTREME_SCALE_URL` may be not set (see the assert on [L24/L30](https://github.com/strongloop/loopback-connector-kv-extreme-scale/pull/12/files#diff-8e5b426819f02c14155e847c4e36d632L24) below).

```js
if (url && /\/wxsdata\/v1\/grids\/[^\/]+\/default$/.test(url))
```

bajtos reviewed (changes requested):
Two more comments, the rest LGTM 👍 . No further review is needed.
